### ISSUE TITLE
Confusion about how to best use touch/mouse functions (coming from Processing)

### ISSUE THREAD

**Issue by JobLeonard** [processing/p5.js#991]
EDIT: The issue described is more-or-less solved by the workaround I posted later, but this still highlights a difference between Processing and p5js, I suspect something to do with JS's asynchronous nature. So until someone chimes in on that I'll leave the issue open.

Not sure if this is the best way to describe the bug, or if this is something p5js can deal with, or if it's purely a hardware/browser issue, but I figured I'd report it anyway:

```
function setup() {
  // uncomment this line to make the canvas the full size of the window
  createCanvas(windowWidth, windowHeight);
  strokeWeight(5);
  stroke(0);
  fill(0);
  background(255);
}

function draw() {
  // draw stuff here
}


function touchMoved(){
  stroke(255, 0, 0);
  line(touchX, touchY, ptouchX, ptouchY);
  return false;
}

function touchEnded(){
  stroke(0, 0, 255);
  line(touchX, touchY, ptouchX, ptouchY);
  return false;
}
```

Try for yourself here:

http://blindedcyclops.neocities.org/draw_bug/index.html

Unless I'm misunderstanding how this is supposed to work, I should see a bunch of disconnected strokes. iOs Safari seems to handle this mostly ok-ish (no screenshot, sorry).

On iOs Google Chrome and desktop Chromium this is what comes out instead:

![image](https://cloud.githubusercontent.com/assets/259840/10393748/da4764aa-6e92-11e5-914c-bc38f7ab8097.png)

If I push down, wait, and then start dragging, I get a proper disconnected line. If I immediately start swiping across the screen, I get the connecting lines that shouldn't be there.

The blue lines are brief taps - strokes never trigger touchEnded apparently.

Firefox Developer Edition on my desktop:

![image](https://cloud.githubusercontent.com/assets/259840/10394003/78edc288-6e94-11e5-9c13-8f2c3e3180a0.png)

Android Chrome on a 2012 Nexus 7 works *almost* like you expect, but the line segments don't perfectly connect because ptouch isn't always updated if I swipe really fast. They do all end with a blue last line segment though.

Android Firefox on the same hardware has the same issue, but even more ptouch lag.

Hope this is useful feedback. Keep up the good work!

**Comment by JobLeonard** [comment]
Update: I added

```
function mouseDragged(){
  stroke(0, 255, 0);
  line(mouseX, mouseY, pmouseX, pmouseY);
}
```
.. to see how mixed input is handled. Basically, pmouse and ptouch are treated as the same variable.

**Comment by JobLeonard** [comment]
Related question, if this is the right place to ask: in Processing, mouseX/mouseY are different variables depending on whether they are used inside draw() or in the mouse functions. I guess that doesn't apply to JavaScript, so what is considered the best practice here?

**Comment by KouichiMatsuda** [comment]
It might be "pixel scaling" working.
You should add  devicePixelScaling(false);

function setup() {
 devicePixelScaling(false);
  // uncomment this line to make the canvas the full size of the window
  createCanvas(windowWidth, windowHeight);
  strokeWeight(5);

**Comment by JobLeonard** [comment]
UPDATE: Semi-answering my own last question, this seems to be a good workaround all browsers I tested before:
```
var ptouchIsDown;
var pmouseIsPressed;
function setup() {
  // uncomment this line to make the canvas the full size of the window
  createCanvas(windowWidth, windowHeight);
  strokeWeight(5);
  stroke(0);
  fill(0);
  background(255);
  ptouchIsDown = touchIsDown;
  pmouseIsPressed = mouseIsPressed;
}

function draw() {
  if (ptouchIsDown && touchIsDown){
    stroke(255, 0, 0);
    line(touchX, touchY, ptouchX, ptouchY);
  }
  if (pmouseIsPressed && mouseIsPressed){
    stroke(0, 0, 255);
    line(mouseX, mouseY, pmouseX, pmouseY);
  }
  ptouchIsDown = touchIsDown;
  pmouseIsPressed = mouseIsPressed;
}

function touchMoved(){
  //stroke(255, 0, 0);
  //line(touchX, touchY, ptouchX, ptouchY);
  return false;
}

function touchEnded(){
  //stroke(0, 0, 255);
  //line(touchX, touchY, ptouchX, ptouchY);
  return false;
}

function mouseDragged(){
  //stroke(0, 255, 0);
  //line(mouseX, mouseY, pmouseX, pmouseY);
}
```
http://blindedcyclops.neocities.org/draw_desktopFix/index.html

From this we also learn that Ubuntu Desktop Firefox doesn't distinguish between mouse and touch, whereas Chrome does.

Not sure if I should close this, or leave it open; might lead to a useful "gotcha for people coming from Processing"-discussion.

**Comment by JobLeonard** [comment]
@KouichiMatsuda: thanks for the suggestion, I've checked if it makes a difference. I've had issues with pixel scaling in the past (especially when playing with the pixels[] array) but it doesn't seem to be the problem here though.

**Comment by limzykenneth** [comment]
I haven't got the right devices to test your problem here but have you tried using mouse events such as `mouseDragged()` instead of touch events. The mouse and touch events are all converted natively by the browsers and probably could give you the intended results.

**Comment by JobLeonard** [comment]
@limzykenneth I've tried both. The issue seems to be one of JavaScript using the same variables for touchX, mouseX, etc. between draw() and the mouse/touch functions. This is different from Processing where they are distinct local variables. So I suspect that that, combined with the asynchronous nature of JavaScript, is what leads to this result.

The code last piece of code I wrote above, which puts everything in the draw() function, is a workaround that solves the problem as far as my use-case is concerned.
